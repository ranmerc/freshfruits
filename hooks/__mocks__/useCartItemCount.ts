const useCartItemCount = jest.fn();

export default useCartItemCount;
