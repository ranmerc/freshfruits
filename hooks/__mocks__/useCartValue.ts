const useCartValue = jest.fn();

export default useCartValue;
