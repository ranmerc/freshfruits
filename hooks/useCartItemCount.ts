import { useCartContext } from "@/context/CartContext";

export default function useCartItemCount() {
  const { cartItems } = useCartContext();

  return cartItems.reduce((sum, item) => {
    return sum + item.quantity;
  }, 0);
}
