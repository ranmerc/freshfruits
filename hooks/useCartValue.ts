import { useCartContext } from "@/context/CartContext";

export default function useCartValue() {
  const { cartItems } = useCartContext();

  return cartItems.reduce((sum, item) => {
    return sum + item.pack.discountPrice * item.quantity;
  }, 0);
}
