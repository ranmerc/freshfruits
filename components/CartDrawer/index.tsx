import { Drawer, NoSsr, Stack, Typography } from "@mui/material";
import Link from "next/link";
import { ButtonBase, Box, Divider } from "@mui/material";
import { ShoppingCart } from "@mui/icons-material";
import useCartValue from "@/hooks/useCartValue";
import useCartItemCount from "@/hooks/useCartItemCount";

export default function CartDrawer() {
  const cartItemCount = useCartItemCount();
  const cartValue = useCartValue();

  if (cartItemCount <= 0) {
    return null;
  }

  return (
    <>
      <NoSsr>
        <Drawer
          variant="permanent"
          anchor="bottom"
          sx={{
            display: { xs: "block", md: "none" },
          }}
        >
          <Box p={2}>
            <Link
              href="/cart"
              style={{ textDecoration: "none", color: "inherit" }}
            >
              <ButtonBase
                sx={{
                  backgroundColor: "secondary.main",
                  color: "secondary.contrastText",
                  padding: "0.8rem 1.5rem",
                  borderRadius: "0.5rem",
                  width: "100%",
                }}
              >
                <Stack
                  direction={"row"}
                  flexWrap={"wrap"}
                  width="100%"
                  justifyContent={"space-between"}
                >
                  <Stack direction={"row"} spacing={1}>
                    <Typography variant="body1">
                      {cartItemCount} items
                    </Typography>
                    <Divider orientation="vertical" variant="middle" flexItem />
                    <Typography variant="body1">â‚¹{cartValue}</Typography>
                  </Stack>
                  <Stack direction={"row"} spacing={0.5}>
                    <ShoppingCart />
                    <Typography>View Cart</Typography>
                  </Stack>
                </Stack>
              </ButtonBase>
            </Link>
          </Box>
        </Drawer>
      </NoSsr>
    </>
  );
}
