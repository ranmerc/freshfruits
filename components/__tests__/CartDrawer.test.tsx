import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import CartDrawer from "../CartDrawer";
import useCartItemCount from "@/hooks/useCartItemCount";
import useCartValue from "@/hooks/useCartValue";
import renderer from "react-test-renderer";

jest.mock("@/hooks/useCartItemCount");
jest.mock("@/hooks/useCartValue");

const mockedUseCartItemCount = useCartItemCount as jest.Mock;
const mockedUseCartValue = useCartValue as jest.Mock;

describe("Cart Drawer Component", () => {
  it("Does not render when cart is empty", () => {
    mockedUseCartItemCount.mockReturnValue(0);
    mockedUseCartValue.mockReturnValue(0);

    render(<CartDrawer />);

    expect(screen.queryByRole("link")).not.toBeInTheDocument();

    const tree = renderer.create(<CartDrawer />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("Renders correctly when cart is not empty", () => {
    mockedUseCartItemCount.mockReturnValue(3);
    mockedUseCartValue.mockReturnValue(90);

    const tree = renderer.create(<CartDrawer />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("Renders SVG logo", () => {
    mockedUseCartItemCount.mockReturnValue(3);
    mockedUseCartValue.mockReturnValue(90);

    render(<CartDrawer />);

    expect(screen.getByTestId("ShoppingCartIcon")).toBeInTheDocument();
  });

  it("Renders cart text", () => {
    mockedUseCartItemCount.mockReturnValue(3);
    mockedUseCartValue.mockReturnValue(90);

    render(<CartDrawer />);

    expect(screen.getByText(/view cart/i)).toBeInTheDocument();
  });

  it("Renders correct item count", () => {
    mockedUseCartItemCount.mockReturnValue(3);
    mockedUseCartValue.mockReturnValue(90);

    render(<CartDrawer />);

    expect(screen.getByText(/3 items/i)).toBeInTheDocument();
  });

  it("Renders correct cart value", () => {
    mockedUseCartItemCount.mockReturnValue(3);
    mockedUseCartValue.mockReturnValue(90);

    render(<CartDrawer />);
    expect(screen.getByText(/â‚¹90/i)).toBeInTheDocument();
  });

  it("Renders link to cart page", () => {
    mockedUseCartItemCount.mockReturnValue(3);
    mockedUseCartValue.mockReturnValue(90);

    render(<CartDrawer />);

    expect(screen.getByRole("link")).toHaveAttribute("href", "/cart");
  });
});
