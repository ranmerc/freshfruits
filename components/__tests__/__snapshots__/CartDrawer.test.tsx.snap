// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Drawer Component Does not render when cart is empty 1`] = `null`;

exports[`Cart Drawer Component Renders correctly when cart is not empty 1`] = `
<div
  className="MuiDrawer-root MuiDrawer-docked css-1ybh3ib-MuiDrawer-docked"
>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiDrawer-paper MuiDrawer-paperAnchorBottom MuiDrawer-paperAnchorDockedBottom css-1ic0da0-MuiPaper-root-MuiDrawer-paper"
  >
    <div
      className="MuiBox-root css-19midj6"
    >
      <a
        href="/cart"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
        style={
          {
            "color": "inherit",
            "textDecoration": "none",
          }
        }
      >
        <button
          className="MuiButtonBase-root css-yb6dg2-MuiButtonBase-root"
          disabled={false}
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <div
            className="MuiStack-root css-1afutji-MuiStack-root"
          >
            <div
              className="MuiStack-root css-niqf4j-MuiStack-root"
            >
              <p
                className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                3
                 items
              </p>
              <hr
                className="MuiDivider-root MuiDivider-middle MuiDivider-vertical MuiDivider-flexItem css-16bghm4-MuiDivider-root"
              />
              <p
                className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                â‚¹
                90
              </p>
            </div>
            <div
              className="MuiStack-root css-69ywkl-MuiStack-root"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ShoppingCartIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zM1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2H1zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2z"
                />
              </svg>
              <p
                className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                View Cart
              </p>
            </div>
          </div>
        </button>
      </a>
    </div>
  </div>
</div>
`;
